src/common/lock.c

```diff
@@ -79,16 +92,34 @@ lockAcquireFile(const String *lockFile, TimeMSec lockTimeout, bool failOnNoLock)
                     // Save the error for reporting outside the loop
                     errNo = errno;
 
-                    // Close the file and reset the file descriptor
+                    // Even though we were unable to lock the file, it may be that it is already locked by another process with the
+                    // same exec-id, i.e. spawned by the same original main process. If so, report the lock as successful.
+                    char buffer[LOCK_BUFFER_SIZE];
+
+                    // Read from file
+                    ssize_t actualBytes = read(result, buffer, sizeof(buffer));
+
+                    // Close the file
                     close(result);
-                    result = -1;
+
+                    // Make sure the read was successful. The file is already open and the chance of a failed read seems pretty
+                    // remote so don't integrate this with the rest of the lock error handling.
+                    THROW_ON_SYS_ERROR_FMT(actualBytes == -1, FileReadError, "unable to read '%s", strZ(lockFile));
+
+                    // Parse the file and see if the exec id matches
+                    const StringList *parse = strLstNewSplitZ(strNewN(buffer, (size_t)actualBytes), LF_Z);
+
+                    if (strLstSize(parse) == 3 && strEq(strLstGet(parse, 1), execId))
+                        result = LOCK_ON_EXEC_ID;
+                    else
+                        result = LOCK_ERROR;
                 }
             }
         }
-        while (result == -1 && (waitMore(wait) || retry));
+        while (result == LOCK_ERROR && (waitMore(wait) || retry));
```

Comment:
storagePathCreate() will throw an error if it fails, so I don't think you can get into an infinite loop here.

You are right that this could be structured a bit differently. You are welcome to create a card for it but I don't want to get diverted right now.

Reply:
Created a "fix" by adding `retry = false` just before the flock check, see this commit f4e93b5d93088c30c5212d1d83fbbae61ef6e7e7 in my branch https://github.com/cmwshang/pgbackrest/tree/cs-dev-lock-ci - we can add it at a later date, so resolving this conversation in this PR.